lib/rmg/RmgArea.cpp

```diff
@@ -0,0 +1,395 @@
+/*
+ * RmgArea.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#include "RmgArea.h"
+#include "CMapGenerator.h"
+
+namespace rmg
+{
+
+void toAbsolute(Tileset & tiles, const int3 & position)
+{
+	Tileset temp;
+	for(auto & tile : tiles)
+	{
+		temp.insert(tile + position);
+	}
+	tiles = std::move(temp);
+}
+
+void toRelative(Tileset & tiles, const int3 & position)
+{
+	toAbsolute(tiles, -position);
+}
+
+Area::Area(const Area & area): dTiles(area.dTiles), dTotalShiftCache(area.dTotalShiftCache)
+{
+}
+
+Area::Area(const Area && area): dTiles(std::move(area.dTiles)), dTotalShiftCache(std::move(area.dTotalShiftCache))
+{
+}
+
+
+Area & Area::operator=(const Area & area)
+{
+	clear();
+	dTiles = area.dTiles;
+	dTotalShiftCache = area.dTotalShiftCache;
+	return *this;
+}
+
+Area::Area(const Tileset & tiles): dTiles(tiles)
+{
+}
+
+Area::Area(const Tileset & relative, const int3 & position): dTiles(relative), dTotalShiftCache(position)
+{
+}
+
+void Area::invalidate()
+{
+	getTiles();
+	dTilesVectorCache.clear();
+	dBorderCache.clear();
+	dBorderOutsideCache.clear();
+}
+
+bool Area::connected() const
+{
+	std::list<int3> queue({*dTiles.begin()});
+	Tileset connected = dTiles; //use invalidated cache - ok
+	while(!queue.empty())
+	{
+		auto t = queue.front();
+		connected.erase(t);
+		queue.pop_front();
+		
+		for(auto & i : int3::getDirs())
+		{
+			if(connected.count(t + i))
+			{
+				queue.push_back(t + i);
+			}
+		}
+	}
+	
+	return connected.empty();
+}
+
+std::list<Area> connectedAreas(const Area & area)
+{
+	std::list<Area> result;
+	Tileset connected = area.getTiles();
+	while(!connected.empty())
+	{
+		result.emplace_back();
+		std::list<int3> queue({*connected.begin()});
+		std::set<int3> queueSet({*connected.begin()});
+		while(!queue.empty())
+		{
+			auto t = queue.front();
+			connected.erase(t);
+			result.back().add(t);
+			queue.pop_front();
+			
+			for(auto & i : int3::getDirs())
+			{
+				auto tile = t + i;
+				if(!queueSet.count(tile) && connected.count(tile) && !result.back().contains(tile))
+				{
+					queueSet.insert(tile);
+					queue.push_back(tile);
+				}
+			}
+		}
+	}
+	return result;
+}
+
+const Tileset & Area::getTiles() const
+{
+	if(dTotalShiftCache != int3())
+	{
+		toAbsolute(dTiles, dTotalShiftCache);
+		dTotalShiftCache = int3();
+	}
+	return dTiles;
+}
+
+const std::vector<int3> & Area::getTilesVector() const
```

Comment:
If we getTiles while dTotalShiftCache is not int3() we refresh tiles and retrieve their correct values. But when we getTilesVector we do not check dTotalShiftCache  and thus can get old data not shifted. This might be hard to trace back taking into account that getTilesVector can be called from various functions like overlap()

Reply:
That was the idea - do not refresh set while translating to avoid multiple set sorting. So getTilesVector should be "life updated" while getTiles - lazy cache mechanism. If you modify area - you invalidate getTilesVector. If you use translation only - you invalidate getTiles. That's the fastest strategy I was able to do. But maybe I understood you point incorrectly

Reply:
Technically it seams to me that translated set should still be in sorted order without additional sorting. So if we can update std::set values and avoid sorting itself this might be an option as well. Now it might be quite ease to misuse these methods because they look quite the same but have a little difference in behavior.

Reply:
I understood. Checked how translate() works.

Reply:
>  translated set should still be in sorted order without additional sorting

I don't know how to do it. Each time when you modify element in set, you will have resorting. In the end you will have same order, but during modifications there will be a lot of extra operations